keps/sig-api-machinery/2839-in-use-protection/README.md

```diff
@@ -0,0 +1,766 @@
+  - Protect `PersistentVolumeClaim` while being used by `Pod` (Replace the background logic for pvc-protection)
+  - Protect `Secret` while being used by `Pod`, `PersistentVolume`, `VolumeSnapshotContent` (Secret protection above)
+  - Protect `ConfigMap` while being used by `Pod`
+  - Protect `StorageClass` while being used by `PersistentVolume`
+  - Protect `VolumeSnapshot` while being used by `VolumeSnapshotContent`
+  - Protect `VolumeSnapshotClass` while being used by `VolumeSnapshotContent`
+  - Protect `User`, `Group`, `ServiceAccount` and `Role` from `RoleBinding`
+  - Protect `User`, `Group`, `ServiceAccount` and `ClusterRole` from `ClusterRoleBinding`
+  - Protect dependent resources while owner resources aren't request to be deleted
+- To avoid accidental deletion of important resources:
+  - Protect resources that controllers or users marked as important
+
+### Goals
+
+- Provide a generic mechanism that prevents a resource, including CRD, from being deleted when it shouldn't be deleted
+- Implement the feature as an advisory feature
+
+### Non-Goals
+
+- Provide a specific mechanism to decide which particular objects should be prevented from being deleted when other particular objects exist
+- Provide a generic mechanism that prevents a resource from being updated when it shouldn't be updated
+- Implement any of the above potential use cases
+
+## Proposal
+
+A new field `Liens` to mark the object not to be deleted is introduced in object Metadata.
+Deletion requests for resources with non-empty `Liens` will be blocked by a newly introduced validation in api-server.
+
+### User Stories (Optional)
+
+#### Story 1
+
+Another controller to protect Secret from deletion watches all Secrets and their potential user objects, like `Pod`, `PersistentVolumes`, and `VolumeSnapshotContent`. Once it find the Secret is used by one of the objects, it updates the Secret's `Liens` and ask this mechanism to block deletion request for the secret.
+
+#### Story 2
+
+A user knows this is an important object, and that terraform likes to delete and recreate objects, so the user sets a lien so that an accidental deletion will fail.
+
+<!--
+### Notes/Constraints/Caveats (Optional)
+-->
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+Most of the design ideas come from [here](https://github.com/kubernetes/kubernetes/issues/10179).
+Users or controllers can add a string to `Liens` to ask to protect a resource.
+A newly introduced validation in api-server will reject deletion requests for resources with non-empty `Liens` and return "423 Locked" error code.
```

Comment:
Using a WebDAV status code feels wrong, especially given that 423 responses
> SHOULD contain an appropriate precondition or postcondition code, such as `lock-token-submitted` or `no-conflicting-lock`.

How about 400, 403, or 409?

Reply:
I don't mind 423, but 409 seems OK: "The request could not be completed due to a conflict with the current state of the resource. This code is only allowed in situations where it is expected that the user might be able to resolve the conflict and resubmit the request."

Reply:
[Precondition/Postcondition XML Elements](https://datatracker.ietf.org/doc/html/rfc4918#section-16) for [423 Locked](https://datatracker.ietf.org/doc/html/rfc4918#section-11.3) looks a bit complex definition to use for this purpose. It is just a recommendation that is defined as "SHOULD" in RFC, so we will be able to ignore it, but we would be able to choose "409 Conflict" which doesn't have such recommendation and it would be good enough. So, I just changed to 409.

Reply:
409 is typically used for retry requests during race conditions, right? Would this prompt controllers to keep retrying the deletion at an incorrect backoff due to the assumption that it is just a concurrent write?

Reply:
That's a valid concern @SOF3. 400 might be a better choice then, as the request is neither forbidden using access controls, conflicted (in the sense that another client's update has caused one) nor locked in the sense that WebDAV would mean it.

If we want to, we can include a message within the 400 response to explain why the request was rejected. Right?